    Q101. Write the Python program to add, subtract, multiply and divide 
    two pandas series ?
        Sample Series- [2, 4, 6, 8, 10], [1, 3, 5, 7, 9]

In [1]:
import pandas as pd
import numpy as np

In [7]:
ser1=pd.Series([2, 4, 6, 8, 10])
ser2=pd.Series([1, 3, 5, 7, 9])
ds = ser1 + ser2
print("Add two Series:")
print(ds)
print("Subtract two Series:")
ds = ser1 - ser2
print(ds)
print("Multiply two Series:")
ds = ser1 * ser2
print(ds)
print("Divide Series1 by Series2:")
ds = ser1 / ser2
print(ds)


Add two Series:
0     3
1     7
2    11
3    15
4    19
dtype: int64
Subtract two Series:
0    1
1    1
2    1
3    1
4    1
dtype: int64
Multiply two Series:
0     2
1    12
2    30
3    56
4    90
dtype: int64
Divide Series1 by Series2:
0    2.000000
1    1.333333
2    1.200000
3    1.142857
4    1.111111
dtype: float64


    Q102. Write a Python program to convert a dictionary to the Pandas 
          Series?
          Sample SeriesOriginal dictionary-
          {'a': 100, 'b': 200, 'c': 300, 'd': 400, 'e': 800}


In [9]:
dict={'a': 100, 'b': 200, 'c': 300, 'd': 400, 'e': 800}
pd_dict=pd.Series({'a': 100, 'b': 200, 'c': 300, 'd': 400, 'e': 800})
pd_dict

a    100
b    200
c    300
d    400
e    800
dtype: int64

    Q103. Write a python program to change the data type of given a column or a Series?

In [10]:
s1 = pd.Series(['100', '200', 'python', '300.12', '400'])
print("Original Series  :" )
print(s1)
conv_ser = pd.to_numeric(s1, errors='coerce')
print("Converted Series  :" )
print(conv_ser)

Original Series  :
0       100
1       200
2    python
3    300.12
4       400
dtype: object
Converted Series  :
0    100.00
1    200.00
2       NaN
3    300.12
4    400.00
dtype: float64


    Q104. Write the python pandas program to convert the first column of a DataFrame as a Series?

In [24]:
df={'col1':[1,2,3,4,7,11],'col2':[4,5,6,9,5,0],'col3':[7,5,8,12,1,11]}
df
dframe=pd.DataFrame(data=df)
s=dframe.iloc[:,0]
print(s)
type(s)

;;;;;;; col1    4.666667
col2    4.833333
col3    7.333333
dtype: float64
0     1
1     2
2     3
3     4
4     7
5    11
Name: col1, dtype: int64


pandas.core.series.Series

    Q105. Write a pandas program to create the mean and standard deviation of the data of a given Series?

In [27]:
s=pd.Series(data=[1,2,3,4,5,6,7,8,9,5,3])
print("Original Data Series:")
print(s)
print("Mean of the Series : ", s.mean())
print("Standard Deviation of the Series : ", s.std())

Original Data Series:
0     1
1     2
2     3
3     4
4     5
5     6
6     7
7     8
8     9
9     5
10    3
dtype: int64
Mean of the Series :  4.818181818181818
Standard Deviation of the Series :  2.522624895547565


    Q106. Write a pandas program to get powers of an array values element-wise?

In [29]:
df = pd.DataFrame(data={'X':[78,85,96,80,86], 'Y':[84,94,89,83,86],'Z':[86,97,96,72,83]});
print(df)

    X   Y   Z
0  78  84  86
1  85  94  97
2  96  89  96
3  80  83  72
4  86  86  83


    Q107. Write the pandas program to get the first 3 rows of a given DataFrame?

In [42]:
exam_data={'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
                          'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
                          'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
                          'qualify': [ 'yes', 'no', 'yes' , 'no', ' no ', ' yes ', 'yes', 'no', 'no', 'yes' ] }
labels=['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j' ]
df=pd.DataFrame(data =exam_data,index=labels)
print("Original data frames  : " )
print(df)
print("First three rows of the data frame:")
print(df.iloc[:3])

Original data frames  : 
        name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James    NaN         3      no
e      Emily    9.0         2     no 
f    Michael   20.0         3    yes 
g    Matthew   14.5         1     yes
h      Laura    NaN         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes
First three rows of the data frame:
        name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes


    Q108: Write the pandas program to select the specified columns and the rows from a given data frame?

In [45]:
print(df.loc[['b','d','f','g'],['name','score']])

      name  score
b     Dima    9.0
d    James    NaN
f  Michael   20.0
g  Matthew   14.5


    Q109. Write the pandas program to calculate mean score for each different student in DataFrame?

In [47]:
print(df['score'].mean())

13.5625


    Q110. Write the Pandas program to rename columns of a given DataFrame ?

In [59]:
data1={'col1': [1,2,3],'col2':[4,5,6],'col2':[7,8,9]}
df=pd.DataFrame(data1)
column_headings = {'col1':'Column1','col2':'Column2','col3':'Column3'}
dff=df.rename(data1,columns=column_headings)
print(dff)

   Column1  Column2
0        1        7
1        2        8
2        3        9


    Q111. Write a pandas program to count city-wise number of people from a given of data set (city, name of the person)?

In [79]:
df1 = pd.DataFrame({'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'city': ['California', 'Los Angeles', 'California', 'California', 'California', 'Los Angeles', 'Los Angeles', 'Georgia', 'Georgia', 'Los Angeles']})

byCity=df1.groupby('city')
print(byCity.size().reset_index(name='Number of people'))

          city  Number of people
0   California                 4
1      Georgia                 2
2  Los Angeles                 4


    Q112. Write a pandas program to widen output display to see more columns?

In [9]:
d = {'col1': [1, 4, 3, 4, 5], 'col2': [4, 5, 6, 7, 8], 'col3': [7, 8, 9, 0, 1]}
df = pd.DataFrame(data=d)
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)
print("Original DataFrame")
print(df)

Original DataFrame
   col1  col2  col3
0     1     4     7
1     4     5     8
2     3     6     9
3     4     7     0
4     5     8     1


    Q113. Write a pandas program to convert the data frame column type from string to DateTime?

In [15]:
s = pd.Series(['3/11/2000', '3/12/2000', '3/13/2000'])
print("String Date:")
print(s)
r=pd.to_datetime(s)
df = pd.DataFrame(r)
print("Original DataFrame (string to datetime):")
print(df)

String Date:
0    3/11/2000
1    3/12/2000
2    3/13/2000
dtype: object
Original DataFrame (string to datetime):
           0
0 2000-03-11
1 2000-03-12
2 2000-03-13


    Q114. Write a pandas program to append the data to an empty DataFrame?

In [18]:
df = pd.DataFrame()
data=pd.DataFrame({'col1':range(3),'col2':range(3),'col3':range(3)})
print("After appending some data:")
df = df.append(data)
print(df)

After appending some data:
   col1  col2  col3
0     0     0     0
1     1     1     1
2     2     2     2


    Q115. Write a pandas program to count the number of columns of a DataFrame?

In [21]:
d = {'col1': [1, 2, 3, 4, 7], 'col2': [4, 5, 6, 9, 5], 'col3': [7, 8, 12, 1, 11]}
df = pd.DataFrame(data=d)
print("Original DataFrame")
print(df)
print("\nNumber of columns:")
print(len(df.columns))


Original DataFrame
   col1  col2  col3
0     1     4     7
1     2     5     8
2     3     6    12
3     4     9     1
4     7     5    11

Number of columns:
3


    Q116. Write a Pandas program to remove the last n rows of a given DataFrame ?

In [24]:
d = {'col1': [1, 2, 3, 4, 7, 11], 'col2': [4, 5, 6, 9, 5, 0], 'col3': [7, 5, 8, 12, 1,11]}
df = pd.DataFrame(data=d)
print("Original DataFrame")
print(df)
print("\nAfter removing last 3 rows of the said DataFrame:")
df1 = df.iloc[:3]
print(df1)

Original DataFrame
   col1  col2  col3
0     1     4     7
1     2     5     5
2     3     6     8
3     4     9    12
4     7     5     1
5    11     0    11

After removing last 3 rows of the said DataFrame:
   col1  col2  col3
0     1     4     7
1     2     5     5
2     3     6     8


    Q117. Write a Pandas program to import excel data (coalpublic2013.xlsx ) into a Pandas data frame.
                         

In [34]:
pd.options.display.max_rows = 10
df=pd.read_excel('resources/coalpublic2013.xlsx')
df

Year  MSHA ID                       Mine_Name  Production  Labor_Hours
0   2013   103381            Tacoa Highwall Miner       56004        22392
1   2013   103404                Reid School Mine       28807        28447
2   2013   100759  North River #1 Underground Min     1440115       474784
3   2013   103246                      Bear Creek       87587        29193
4   2013   103451                     Knight Mine      147499        46393
..   ...      ...                             ...         ...          ...
45  2013  1519322                         Ghm #25       25054         3108
46  2013   103321                  Poplar Springs      189370        76366
47  2013   103358                       Old Union      284563       161805
48  2013  5000030                        Usibelli     1631584       286079
49  2013   201195                    Kayenta Mine     7602722      1015333

[50 rows x 5 columns]

    Q118. Write a Pandas program to import excel data (coalpublic2013.xlsx ) into a data frame and find details where "Mine Name" starts with "P.

In [42]:
df[df['Mine_Name'].map(lambda x : x.startswith("P"))].head()

Year  MSHA ID                      Mine_Name  Production  Labor_Hours
13  2013   103332                  Powhatan Mine      140521        61394
18  2013   102976  Piney Woods Preparation Plant           0        14828
19  2013   102976  Piney Woods Preparation Plant           0        23193
46  2013   103321                 Poplar Springs      189370        76366

    Q119. Write a Pandas program to import excel data (employee.xlsx )into a Pandas dataframe and find the list of employees where hire_date> 01-01-07.

In [47]:
empdf=pd.read_excel('resources/employee.xlsx')
empdf[empdf['hire_date'] >='20070101']

emp_id first_name   last_name  hire_date
4      104      Bruce       Ernst 2007-05-21
7      107      Diana     Lorentz 2007-02-07
13     113       Luis        Popp 2007-12-07
19     119      Karen  Colmenares 2007-08-10

    Q120. Write a Pandas program to import excel data (employee.xlsx ) into a Pandas dataframe and find a list of the employees of  a specified year

In [52]:
empdf2 = empdf.set_index(['hire_date'])
resultempd = empdf2["2005"]
resultempd

emp_id first_name last_name
hire_date                              
2005-09-21     101      Neena   Kochhar
2005-06-25     105      David    Austin
2005-09-28     110       John      Chen
2005-09-30     111     Ismael   Sciarra
2005-12-24     116     Shelli     Baida
2005-07-24     117      Sigal    Tobias

    Q121. Write a pandas program to import three datasheets from a given excel data (coalpublic2013.xlsx ) in to a single dataframe.

In [7]:
df1=pd.read_excel('resources/employee.xlsx',sheet_name=0)
df2=pd.read_excel('resources/employee.xlsx',sheet_name=1)
df3=pd.read_excel('resources/employee.xlsx',sheet_name=2)
df_total=pd.concat([df1,df2,df3])
df_total.reset_index(drop=True, inplace=True)
df_total

emp_id   first_name    last_name  hire_date
0      100       Steven         King 2003-06-17
1      101        Neena      Kochhar 2005-09-21
2      102          Lex      De Haan 2001-01-13
3      103    Alexander       Hunold 2006-01-03
4      104        Bruce        Ernst 2007-05-21
5      105        David       Austin 2005-06-25
6      106        Valli    Pataballa 2006-02-05
7      107        Diana      Lorentz 2007-02-07
8      108        Nancy    Greenberg 2002-08-17
9      109       Daniel       Faviet 2002-08-16
10     110         John         Chen 2005-09-28
11     111       Ismael      Sciarra 2005-09-30
12     112  Jose Manuel        Urman 2006-03-07
13     113         Luis         Popp 2007-12-07
14     114          Den     Raphaely 2002-12-07
15     115    Alexander         Khoo 2003-05-18
16     116       Shelli        Baida 2005-12-24
17     117        Sigal       Tobias 2005-07-24
18     118          Guy       Himuro 2006-11-15
19     119        Karen   Colmenares 2007-08-10
20     120      Matthew        Weiss 2004-07-18
21     121         Adam        Fripp 2005-04-10
22     122        Payam     Kaufling 2003-05-01
23     123       Shanta      Vollman 2005-10-10
24     124        Kevin      Mourgos 2007-11-16
25     125        Julia        Nayer 2005-07-16
26     126        Irene  Mikkilineni 2006-09-28
27     127        James       Landry 2007-01-14
28     128       Steven       Markle 2008-03-08
29     129        Laura       Bissot 2005-08-20
30     130        Mozhe     Atkinson 2005-10-30
31     131        James       Marlow 2005-02-16
32     132           TJ        Olson 2007-04-10
33     133        Jason       Mallin 2004-06-14
34     134      Michael       Rogers 2006-08-26
35     135           Ki          Gee 2007-12-12
36     136        Hazel   Philtanker 2008-02-06
37     137       Renske       Ladwig 2003-07-14
38     138      Stephen       Stiles 2005-10-26
39     141       Trenna         Rajs 2003-10-17
40     142       Curtis       Davies 2005-01-29
41     143      Randall        Matos 2006-03-15
42     144        Peter       Vargas 2006-07-09
43     145         John      Russell 2004-10-01
44     146        Karen     Partners 2005-01-05
45     147      Alberto    Errazuriz 2005-03-10
46     148       Gerald    Cambrault 2007-10-15
47     149        Eleni      Zlotkey 2008-01-29

    Q 122. Write a pandas program to import three datasheets from a given excel data (employee.xlsx ) into a single data frame and export the result into new Excel file.

In [9]:
df_total.to_excel('resources/new_emp.xlsx', index=False)

    Q123. Write a pandas program to create the Pivot table with multiple indexes from the data set of the titanic.csv.

In [2]:
df = pd.read_csv('resources/titanic.csv')
result = pd.pivot_table(df, index = ["sex","age"], aggfunc=np.sum)
result

Unnamed: 15  adult_male  alone      fare  parch  pclass  sibsp  \
sex    age                                                                     
female 0.75           0.0         0.0    0.0   38.5166      2       6      4   
       1.00           0.0         0.0    0.0   26.8750      3       6      1   
       2.00           0.0         0.0    0.0  259.4750      9      15      9   
       3.00           0.0         0.0    0.0   62.6542      3       5      4   
       4.00           0.0         0.0    0.0  114.1417      6      13      4   
...                   ...         ...    ...       ...    ...     ...    ...   
male   70.00          0.0         2.0    1.0   81.5000      1       3      1   
       70.50          0.0         1.0    1.0    7.7500      0       3      0   
       71.00          0.0         2.0    2.0   84.1584      0       2      0   
       74.00          0.0         1.0    1.0    7.7750      0       3      0   
       80.00          0.0         1.0    1.0   30.0000      0       1      0   

              survived  
sex    age              
female 0.75          2  
       1.00          2  
       2.00          2  
       3.00          1  
       4.00          5  
...                ...  
male   70.00         0  
       70.50         0  
       71.00         0  
       74.00         0  
       80.00         1  

[145 rows x 8 columns]

    Q124. Write a Pandas program to create the Pivot table and find survival rate by gender?

In [44]:
result = df.pivot_table( 'survived',index=['sex'])
result

survived
sex             
female  0.742038
male    0.188908

    Q125. Write a pandas program to make partition each of the passengers into 4 categories based on their age.

In [42]:
result = pd.cut(df['age'], [0, 10, 30, 60, 80])
result

0      (10.0, 30.0]
1      (30.0, 60.0]
2      (10.0, 30.0]
3      (30.0, 60.0]
4      (30.0, 60.0]
           ...     
886    (10.0, 30.0]
887    (10.0, 30.0]
888             NaN
889    (10.0, 30.0]
890    (30.0, 60.0]
Name: age, Length: 891, dtype: category
Categories (4, interval[int64]): [(0, 10] < (10, 30] < (30, 60] < (60, 80]]

    Q126. Write a pandas program to create the Pivot table and find survival rate by the gender, age of the different categories of various classes.

In [41]:
fare = pd.qcut(df['fare'], 2)
age = pd.cut(df['age'], [0, 10, 30, 60, 80])
result = df.pivot_table( 'survived',index=['sex', age], columns=[fare, 'pclass'])
result

fare            (-0.001, 14.454]                     (14.454, 512.329]  \
pclass                         1         2         3                 1   
sex    age                                                               
female (0, 10]               NaN       NaN  0.800000          0.000000   
       (10, 30]              NaN  0.933333  0.568182          0.970588   
       (30, 60]              NaN  0.846154  0.142857          0.979167   
       (60, 80]              NaN       NaN  1.000000          1.000000   
male   (0, 10]               NaN       NaN  1.000000          1.000000   
       (10, 30]              NaN  0.034483  0.140625          0.458333   
       (30, 60]              0.0  0.130435  0.109375          0.440678   
       (60, 80]              NaN  0.333333  0.000000          0.083333   

fare                                 
pclass                  2         3  
sex    age                           
female (0, 10]   1.000000  0.411765  
       (10, 30]  0.904762  0.307692  
       (30, 60]  0.941176  0.333333  
       (60, 80]       NaN       NaN  
male   (0, 10]   1.000000  0.263158  
       (10, 30]  0.000000  0.130435  
       (30, 60]  0.047619  0.166667  
       (60, 80]       NaN       NaN

    Q127. Write a pandas program to create the Pivot table and calculate number of women and men were in a particular cabin class.

In [45]:
result = df.pivot_table(index=['sex'], columns=['pclass'], aggfunc='count')
result

Unnamed: 15       adult_male            age          alive  ... parch  \
pclass           1  2  3          1    2    3    1   2    3     1  ...     3   
sex                                                                ...         
female           0  0  0         94   76  144   85  74  102    94  ...   144   
male             0  0  0        122  108  347  101  99  253   122  ...   347   

       sibsp           survived            who            
pclass     1    2    3        1    2    3    1    2    3  
sex                                                       
female    94   76  144       94   76  144   94   76  144  
male     122  108  347      122  108  347  122  108  347  

[2 rows x 42 columns]

    Q128. Write a pandas program to create the Pivot table and separate the gender according to whether they travelled alone or not to get the probability of survival

In [19]:
df = pd.read_csv('resources/titanic.csv')
result = df.pivot_table('survived', ['sex','alone'],columns='class')
result

class            First    Second     Third
sex    alone                              
female False  0.966667  0.931818  0.416667
       True   0.970588  0.906250  0.616667
male   False  0.425532  0.277778  0.180723
       True   0.333333  0.097222  0.121212

    Q129. Write a pandas program to create the Pivot table and find the probability of survival by class, gender, solo boarding, and the port of embarkation.

In [22]:
df = pd.read_csv('resources/titanic.csv')
result = df.pivot_table('survived', ['sex' , 'alone' ], [ 'embark_town', 'class' ])
result

embark_town  Cherbourg                     Queenstown                   \
class            First    Second     Third      First Second     Third   
sex    alone                                                             
female False  1.000000  1.000000  0.611111        1.0    NaN  0.625000   
       True   0.944444  1.000000  0.800000        NaN    1.0  0.760000   
male   False  0.473684  0.166667  0.500000        0.0    NaN  0.100000   
       True   0.347826  0.250000  0.151515        NaN    0.0  0.068966   

embark_town  Southampton                      
class              First    Second     Third  
sex    alone                                  
female False    0.941176  0.923077  0.327586  
       True     1.000000  0.892857  0.466667  
male   False    0.407407  0.300000  0.142857  
       True     0.326923  0.089552  0.123762

    Q130. Write a pandas program to get current date, oldest date and number of days between Current date and the oldest date of Ufo dataset.